packages/type-utils/src/typeOrValueSpecifiers/typeDeclaredInPackageDeclarationFile.ts

```diff
@@ -0,0 +1,55 @@
+import * as ts from 'typescript';
+
+function findParentModuleDeclaration(
+  node: ts.Node,
+): ts.ModuleDeclaration | undefined {
+  switch (node.kind) {
+    case ts.SyntaxKind.ModuleDeclaration:
+      return node as ts.ModuleDeclaration;
+    case ts.SyntaxKind.SourceFile:
+      return undefined;
+    default:
+      return findParentModuleDeclaration(node.parent);
+  }
+}
+
+function typeDeclaredInDeclareModule(
+  packageName: string,
+  declarations: ts.Node[],
+): boolean {
+  return declarations.some(
+    declaration =>
+      findParentModuleDeclaration(declaration)?.name.text === packageName,
```

Comment:
@JoshuaKGoldberg I was investigating your work in this PR because I'm hoping to use it in #8556.

Here's a question I had:
In the `main` branch, it is impossible to match the built-in type `URL` if `@types/node` is installed ([explanation](https://github.com/typescript-eslint/typescript-eslint/pull/8556#issuecomment-2165608615)).

In this PR, I see that it is now possible to match the built-in type `URL` when `@types/node` is installed.
For example, the specifier `{ from: 'package', name: 'URL', package: 'global' }` matches against `type Test = URL;`.

This is because `findParentModuleDeclaration(declaration)?.name.text` returns `'global'`.

Is this expected? I'm not 100% sure where `global` is coming from — I believe it's a `declare global` somewhere.

Reply:
Huh, that's ... a really interesting bug! Thanks for pointing it out. `findParentModuleDeclaration` should be filtering out `declare global` but isn't. Nice find!

Reply:
(I fixed that in this PR)